### PR TITLE
update object.h definitions for Python 3.12

### PR DESCRIPTION
**Pull request by davidhewitt**
This fixes a flaky reference counting issue observed in tests which seems to be related to the new "immortal objects" of Python 3.12.

So many changes had happened to `object.h` due to the immortal objects I felt it better to do a full reconciliation of the definitions in that file against the 3.12 header.

The reference counting changes are complicated, so I'd like to see a full CI run and also add some more tests before calling this ready to merge.

### CODE REVIEW

File: pyo3-ffi/src/object.rs
Comment (davidhewitt) [reply pyo3-ffi/src/object.rs:46]: Agreed, great catch 👍

**Comment by davidhewitt** [comment]
Ok, looks like we're there 🎉 

I changed tests which interact with reference counting to use `object()` instead of `None`, because now that `None` is immortal on 3.12 it's not a suitable value to use in tests which try to modify reference counts!

**Comment by davidhewitt** [comment]
I'd quite like to merge this soon so that I can proceed with the patch release. The following summarises the diff, to help with reviewing:
- On Python 3.11 `buffer.h` was renamed to `pybuffer.h`, so I made the corresponding rename here.
- On Python 3.12 the definitions of `getbufferproc` and `releasebufferproc` have been moved to `pybuffer.h`. (So I moved them for all versions here.)
- On Python 3.12 the `object.h` file has new symbols `_Py_IMMORTAL_REFCNT` and `_Py_IsImmortal`. The implementation of these differs for 32-bit architectures:
  - On 32-bit architectures, the lower 30 bits of the 32-bit reference count are set to 1 to denote an immortal object.
  - On 64-bit architectures, the lower 32 bits are set instead.
  - I expect this will all change again in future when PEP 703 / nogil comes into the picture with thread-biased reference counting.
- I took the argument names out of typedefs for `unaryfunc` etc, because the Python header doesn't give those arguments names and the argument names were just `arg1` `arg2` rather than anything meaningful. 
- I updated the implementations of a few inline macros in `object.rs`, adding helper definitions to `pyport.rs` where these were used. Most of these are trivial, except for `Py_INCREF` and `Py_DECREF`.
  - I added some tests to `src/ffi/tests.rs` to verify the behaviour of these functions. To exercise these with `abi3` I had to change that file to exclude individual tests with `#[cfg(not(Py_LIMITED_API))]` rather than excluding the whole module.

- Finally, I updated other refcounting tests to use `object()` instead of `None`, as per comment above.

I appreciate this is quite a meaty patch 😢

File: pyo3-ffi/src/object.rs
Comment (adamreichold) [inline pyo3-ffi/src/object.rs:47]: I think this only works for `target_pointer_width = "64"` and would hence would prefer that as the `cfg` attribute here and above. (I don't think we can reasonably CPython on CHERI here in any case.)

File: pyo3-ffi/src/object.rs
Comment (adamreichold) [inline pyo3-ffi/src/object.rs:45]: I suspect the union is anonymous in the C header, but can't we give it a nicer name here, e.g. `PyObjectRefcnt` or something like that?

File: pyo3-ffi/src/object.rs
Comment (adamreichold) [inline pyo3-ffi/src/object.rs:125]: Couldn't this use the split representation? Or is this copied verbatim from the C header?

File: pyo3-ffi/src/object.rs
Comment (adamreichold) [inline pyo3-ffi/src/object.rs:131]: Are the top two bits never modified for immortal objects?

File: pyo3-ffi/src/object.rs
Comment (adamreichold) [inline pyo3-ffi/src/object.rs:383]: Is this intentionally equal to `Py_TPFLAGS_MANAGED_DICT`?

File: pyo3-ffi/src/object.rs
Comment (adamreichold) [inline pyo3-ffi/src/object.rs:518]: Couldn't this part be shared as in the decrement implementation?

**Review by adamreichold — COMMENTED** [review]
I have to say that CPython is certainly not becoming less baroque.

Besides the inline questions on trying to better abstract out the differences in how the reference counts are handled, I wonder whether `PyAny::get_refcnt` would have to change as this looks more like `Option<isize>` then `isize` conceptionally?

File: pyo3-ffi/src/object.rs
Comment (davidhewitt) [reply pyo3-ffi/src/object.rs:45]: Yes it's anonymous in the C header. I quite wanted to give it a name which implied that PyO3 had been forced to invent a name. I'll go for `PyObjectObRefcnt` (as the field is `ob_refcnt`), and add a docstring explaining this type is invented by PyO3.

File: pyo3-ffi/src/object.rs
Comment (davidhewitt) [reply pyo3-ffi/src/object.rs:47]: The C header used `# if SIZEOF_VOID_P > 4`, of which the closest analogy I could find was `not(target_pointer_width = "32")`, disregarding 16 bits which I don't think we would try to support.

File: pyo3-ffi/src/object.rs
Comment (davidhewitt) [reply pyo3-ffi/src/object.rs:125]: Yep copied verbatim: https://github.com/python/cpython/blob/e3b5ed7b1cb8ad12aa285cf78303c6dda0953181/Include/object.h#L243

File: pyo3-ffi/src/object.rs
Comment (davidhewitt) [reply pyo3-ffi/src/object.rs:47]: https://github.com/python/cpython/blob/e3b5ed7b1cb8ad12aa285cf78303c6dda0953181/Include/object.h#L182-L184

File: pyo3-ffi/src/object.rs
Comment (davidhewitt) [reply pyo3-ffi/src/object.rs:131]: This implementation matches the C header: https://github.com/python/cpython/blob/e3b5ed7b1cb8ad12aa285cf78303c6dda0953181/Include/object.h#L245

I think in practice the top two bits are reserved for abi3 builds built against older Python implementations, for which `Py_INCREF` would still directly modify the refcount field and so immortal objects need to allow some space for the count to spill over without wrapping.

File: pyo3-ffi/src/object.rs
Comment (davidhewitt) [reply pyo3-ffi/src/object.rs:383]: `MANAGED_DICT` should be `1 << 4`, great catch 👍

File: pyo3-ffi/src/object.rs
Comment (davidhewitt) [reply pyo3-ffi/src/object.rs:518]: In the header it's split like this where the "immortality" in the 64-bit implementation is achieved by the `wrapping_add` to the lower 32 bits will hit 0, and checking for it.

https://github.com/python/cpython/blob/e3b5ed7b1cb8ad12aa285cf78303c6dda0953181/Include/object.h#L640-L654

I agree it looks a bit funky, presumably there was a technical reason this was preferred. I'd like to match the C implementation as faithfully as possible.

(I'm also hoping that when nogil comes along the whole macro is removed so that we don't have to try to reproduce it here 😂)

**Comment by davidhewitt** [comment]
Pushed a commit with review corrections, I'll squash before merge.

I think there is an open question for `PyAny::get_refcnt`. I'll open an issue.

File: pyo3-ffi/src/object.rs
Comment (adamreichold) [reply pyo3-ffi/src/object.rs:49]: I think copying verbatim is not useful here and would prefer the direct formulation as the structure does not make sense for anything but 64 bit architectures.

File: pyo3-ffi/src/object.rs
Comment (adamreichold) [reply pyo3-ffi/src/object.rs:133]: Ok, the salient point here is probably

> The object would lose its immortality, but the execution would still be correct.

from the C header. So indeed, in 32-bit builds, these object would really not be considered "immortal" any more after such an extension touched their reference.

Weird, but matching upstream behaviour. (I really wonder whether the benefit of having immortal objects is worth this technical complexity.)

**Review by adamreichold — APPROVED** [review]
LGTM modulo the usage of `target_point_width`.

**Comment by davidhewitt** [comment]
🙏 thanks for the thorough reviewing!

Changed to use `cfg(target_pointer_width = "64")` instead of `cfg(not(target_pointer_width = "32"))`. We won't support 16 or 128-bit systems for now, can cross that bridge then it seems like there's a reason to.

**Pull request merged**
